### PR TITLE
[PyTorch] Replace IValue::intrusive_ptr with IValue::isIntrusivePtr()

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76425
* #76418

This tidies up an extra piece of state that was computable from the other state in IValue.

Differential Revision: [D31290017](https://our.internmc.facebook.com/intern/diff/D31290017/)

### CODE REVIEW

File: aten/src/ATen/core/ivalue.h [hunk 1]
```diff
@@ -365,11 +365,14 @@ struct TORCH_API IValue final {
...
 
-    if (!is_intrusive_ptr) {
+    if (!isIntrusivePtr()) {
       return 1;
     }
 
     if (payload.u.as_intrusive_ptr == c10::UndefinedTensorImpl::singleton()) {
+      if (isStorage() || isGenerator()) {
+        return 1;
+      }
```
Comment (peterbell10) [inline aten/src/ATen/core/ivalue.h:375]: This deserves a comment.

File: aten/src/ATen/core/ivalue.h [hunk 2]
```diff
@@ -921,7 +913,13 @@ struct TORCH_API IValue final {
...
 
   bool isPtrType() const {
-    return (isTensor() && payload.as_tensor.defined()) || is_intrusive_ptr;
+    if (isTensor()) {
+      return payload.as_tensor.defined();
+    }
+    if (isStorage() || isGenerator()) {
+      return payload.u.as_intrusive_ptr != c10::UndefinedTensorImpl::singleton();
+    }
+    return isIntrusivePtr();
```
Comment (peterbell10) [inline aten/src/ATen/core/ivalue.h:922]: I presume this is for BC, so wouldn't it make sense to use `isIntrusivePtrLegacyBehavior()`?

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/core/ivalue.h (hunk 2 above)
Comment (swolchok) [reply aten/src/ATen/core/ivalue.h:922]: nice catch!

File: aten/src/ATen/core/ivalue.h (hunk 1 above)
Comment (swolchok) [reply aten/src/ATen/core/ivalue.h:375]: changing to use isIntrusivePtrLegacyBehavior

**Comment by ezyang** [comment]
@JackCaoG XLA may need to track this change

**Review by peterbell10 — APPROVED** [review]

**Comment by swolchok** [comment]
not a believer on the two CI failures

@pytorchbot merge this please

**Pull request closed without merging**